### ISSUE TITLE
Proposal for restructure to support CI tests, pyspec, and better test generation flow.

### ISSUE THREAD

**Issue by protolambda** [ethereum/eth2.0-specs#836]
Proposal to have clean CI integration, with testing, test-generation, and pythonized-spec.

It would look a little bit like a mono-repo, but well-structured, and enabling us to:
- write pytests for CI info during PR, mostly sanity tests, based on pythonized spec.
- have generators use the pythonized spec easily (ramp up amount of yaml tests)
- easily write generators (thanks to base generator from PR 31)
- make yaml test repo builds directly hooked to spec releases, as it should when they are based on the spec. (i.e. BLS/SSZ/state-transition generators can all use existing spec/util code, and do not duplicate it to a generators repo)
- Clean up current CI. It is great to have, but I hate seeing a requirements.txt in the root of a project that is meant to be a specification. Separating into directories and preserving the specs directory to be markdown-only feels like the best I can do to both support "specification", and strong CI + tests generation.

```
├── LICENSE
├── Makefile   <--- see description below.
├── README.md
├── scripts
│   ├── requirements.txt
│   └── phase0
│       ├── build_spec.py
│       ├── function_puller.py
│       └── __init__.py
├── specs
│   ├── bls_signature.md
│   ├── core
│   │   ├── 0_beacon-chain.md
│   │   └── 1_shard-data-chains.md
│   ├── light_client
│   │   ├── merkle_proofs.md
│   │   └── sync_protocol.md
│   ├── simple-serialize.md
│   ├── test-format.md
│   └── validator
│       └── 0_beacon-chain-validator.md
├── test_generators
│   ├── bls
│   ├── ... more ...
│   ├── shuffling
│   └── ssz
├── test_libs
│   ├── gen_helpers
│   │   ├── gen_base <-- package that generators use as base (test gen PR 31)
│   │   │   └── ....
│   │   ├── README.md
│   │   ├── requirements.txt
│   │   └── setup.py
│   └── pyspec
│       ├── requirements.txt
│       ├── README.md
│       ├── setup.py <--- defines the pyspec module. To local-require in generators
│       ├── phase0 <--- package, generated using spec builder script, by makefile
│       ├── phase1,2,etc. <--- later phase packages
│       └── utils <--- shared between phase packages.
│           ├── __init__.py
│           └── phase0 <--- utilities for phase 0,
│                           we may want to generalize utils between phases as well.
├── py_tests  <--- existing py-tests, but not confused with yaml tests
│   ├── requirements.txt
│   ├── README.md
│   └── phase0
│       ├── __init__.py
│       ├── block_processing
│       │   ├── test_process_block_header.py
│       │   ├── test_process_deposit.py
│       │   ├── test_process_proposer_slashing.py
│       │   └── test_voluntary_exit.py
│       ├── conftest.py
│       ├── helpers.py
│       └── test_sanity.py
└── yaml_tests <---- NOT included in git, build output from test generators.
    │                 Output by CI to eth2.0-tests repo.
    ├── bls
    ├── ... more ...
    ├── shuffling
    └── ssz

```

Makefile:
- defines a target for `./test_libs/pyspec/phase0/`. The `./scripts/phase0/build_spec.py` should be called.
- defines a target for `./test_libs/pyspec/`. This calls all pyspec targets.
- defines yaml_test make targets, i.e. test-gen runners. Use new makefile code from [test-gen PR 31](https://github.com/ethereum/eth2.0-test-generators/pull/31). This automatically creates targets for each generator. These targets should have `./test_libs/pyspec/` as dependency.
- defines a `test` target to build and run `py_tests`. This target should define `./test_libs/pyspec/` as dependency.

**Comment by djrtwo** [comment]
My gut is that this is cleaner and going to provide a simpler experience writing and testing the spec. The more repos we have with inter-dependencies, the more complex doing any one of these tasks is going to be.

The separate test generator repo made more sense when we did not yet have an executable spec and were pulling in dependencies from various repos. At this point the executable spec as the entire spec functionality, the shuffling alg, and a minimal ssz/hash-tree-root. We can also use the internal fork choice function with a test_case to `BeaconBlock` converter to generate fork choice tests. The only external dependency required for testing at this point is bls.

I am in favor of this. Curious to hear the thoughts of some of the other people that have been working on testing. cc: @jannikluhn @ChihChengLiang @mratsim @paulhauner

**Comment by ChihChengLiang** [comment]
One downside is the test generator related issues and PRs are coming to this repo too, not sure how these affect the spec contributors. Otherwise, with the executable spec, the restructure looks beautiful.

**Comment by jannikluhn** [comment]
I'm a bit hesitant on this. It would tie releases of test generators to releases of the spec even though we'll probably want to have them independent (hopefully more frequent test releases than spec releases in the future). Right now we build the tests on every release of the generators (which I quite like, it's simple, quick, and transparent), so we would need to find another way for this, in the worst case clicking some button hidden in the CircleCI UI.

I don't really see an advantage of the proposed approach compared to putting the state test generators in the generators repo. We can configure CI in such a way that it pulls the latest spec release whenever building the tests. Then generator could use any spec code they like as well and we're sure that newly generated tests are always up to date.

**Comment by djrtwo** [comment]
> I don't really see an advantage of the proposed approach compared to putting the state test generators in the generators repo.

That main benefit is to have the py-tests live alongside the spec to be able to run CI testing on the spec when writing and updating. If the test generators live in a separate repo, then it is much more complex to get the py-tests we are using on the spec to be cleanly integrated into the generators.  The more we separate the two, the more complex the link, dependencies, and general dev flow has to be, _or_ the py-tests just remain independent of test generation and we have to duplicate testing work to test these common flows.

Attaching a `t0` to spec releases (`v0.5.0-t0`) and incrementing the `t` appendage upon testing updates (with no substantive spec updates) is one way we can hang onto the same release build option. 

> One downside is the test generator related issues and PRs are coming to this repo too, not sure how these affect the spec contributors.

We could have testing related issues live in `eth2.0-tests` to keep a clean separation.

**Comment by djrtwo** [comment]
We're going to move forward with this proposal

**Comment by protolambda** [comment]
I'll start implementing this proposal tomorrow, can't wait to see the spec function as a direct base for test suites

**Comment by djrtwo** [comment]
closed via #851
